src/arc.rs

```diff
@@ -0,0 +1,215 @@
+#![allow(unsafe_code)]
+/// We use this because we never use the weak count on the std
+/// `Arc`, but we use a LOT of `Arc`'s, so the extra 8 bytes
+/// turn into a huge overhead.
+use std::{
+    alloc::{alloc, dealloc, Layout},
+    convert::TryFrom,
+    fmt::{self, Debug},
+    mem,
+    ops::Deref,
+    ptr,
+    sync::atomic::{AtomicUsize, Ordering},
+};
+
+struct ArcInner<T: ?Sized> {
+    rc: AtomicUsize,
+    inner: T,
+}
+
+pub struct Arc<T: ?Sized> {
+    ptr: *mut ArcInner<T>,
+}
+
+unsafe impl<T: Send + Sync + ?Sized> Send for Arc<T> {}
+unsafe impl<T: Send + Sync + ?Sized> Sync for Arc<T> {}
+
+impl<T: Debug> Debug for Arc<T> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> Result<(), fmt::Error> {
+        f.debug_map().entry(&"inner", unsafe { &(*self.ptr).inner }).finish()
+    }
+}
+
+impl<T> Arc<T> {
+    pub fn new(inner: T) -> Arc<T> {
+        let bx = Box::new(ArcInner { inner, rc: AtomicUsize::new(1) });
+        let ptr = Box::into_raw(bx);
+        Arc { ptr }
+    }
+
+    pub fn copy_from_slice(s: &[T]) -> Arc<[T]> {
+        let align =
+            std::cmp::max(mem::align_of::<T>(), mem::align_of::<AtomicUsize>());
+
+        let rc_width = std::cmp::max(align, mem::size_of::<AtomicUsize>());
+        let data_width = mem::size_of::<T>().checked_add(s.len()).unwrap();
```

Comment:
I think this should be a checked_mul here

Reply:
good catch, I think it's currently only used for u8's which is why that part didn't blow up yet